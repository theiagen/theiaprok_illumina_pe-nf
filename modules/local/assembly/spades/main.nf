process SPADES {
    tag "$meta.id"
    label "process_medium"
    
    container "us-docker.pkg.dev/general-theiagen/staphb/spades:4.1.0"

    input:
    tuple val(meta), path(reads)
    val kmers
    val spades_type

    output:
    tuple val(meta), path('*.contigs.fa'), emit: contigs
    tuple val(meta), path('*.contigs.gfa'), emit: gfa, optional: true
    tuple val(meta), path("*.log")        , emit: log
    path "versions.yml"                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def maxmem = task.memory.toGiga()
    def type = spades_type ?: 'isolate'
    def k_mer = kmers ? "-k ${kmers}" : ""
    def reads_input = meta.single_end ? "-s ${reads}" : "-1 ${reads[0]} -2 ${reads[1]}"
    
    """
    spades.py \\
        --${type} \\
        ${reads_input} \\
        ${k_mer} \\
        --threads ${task.cpus} \\
        --memory ${maxmem} \\
        --phred-offset 33 \\
        -o spades \\
        ${args}

    # Handle metaviral special case
    if [[ "${type}" == "metaviral" ]]; then
        if [[ -f spades/contigs.fasta ]]; then
            echo "PASS" > spades_status.txt
            mv spades/contigs.fasta ${prefix}_${type}spades_contigs.fa
            if [[ -f spades/assembly_graph_with_scaffolds.gfa ]]; then
                mv spades/assembly_graph_with_scaffolds.gfa ${prefix}_${type}spades_contigs.gfa
            fi
        else
            echo "FAIL" > spades_status.txt
            touch ${prefix}_${type}spades_contigs.fa
        fi
    else
        echo "PASS" > spades_status.txt
        mv spades/contigs.fasta ${prefix}_${type}spades_contigs.fa
        if [[ -f spades/assembly_graph_with_scaffolds.gfa ]]; then
            mv spades/assembly_graph_with_scaffolds.gfa ${prefix}_${type}spades_contigs.gfa
        fi
    fi

    mv spades/spades.log ${prefix}_spades.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        spades: \$(spades.py --version 2>&1 | sed 's/^.*SPAdes genome assembler v//; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def type = spades_type ?: 'isolate'
    """
    touch ${prefix}_${type}spades_contigs.fa
    touch ${prefix}_${type}spades_contigs.gfa
    touch ${prefix}_spades.log
    echo "PASS" > spades_status.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        spades: \$(spades.py --version 2>&1 | sed 's/^.*SPAdes genome assembler v//; s/ .*\$//')
    END_VERSIONS
    """
}