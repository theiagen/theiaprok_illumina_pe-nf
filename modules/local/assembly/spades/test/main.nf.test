nextflow_process {
    name "Test Process SPADES"
    script "../main.nf"
    process "SPADES"
    
    test("spades - test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = ""
                input[2] = "isolate"
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.assembly_fasta,
                    process.out.assembly_gfa,
                    process.out.assembly_log,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("fastqc - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = '21,33,55'
                input[2] = 'isolate'
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}