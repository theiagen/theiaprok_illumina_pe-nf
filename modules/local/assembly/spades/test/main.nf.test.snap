{
    "fastqc - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_isolatespades_contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_isolatespades_contigs.gfa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_spades.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,07e86bb6f39d315279103919a01645a5"
                ],
                "assembly_fasta": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_isolatespades_contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "assembly_gfa": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_isolatespades_contigs.gfa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "assembly_log": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_spades.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,07e86bb6f39d315279103919a01645a5"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T14:45:04.601207949"
    },
    "spades - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_isolatespades_contigs.fa:md5,6ff16a20c43701b8fbf4afee544dc905"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_isolatespades_contigs.gfa:md5,18d254ebaccd286337f065b028050437"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_spades.log:md5,463343cfc24aa52a559f270888e38c00"
                ]
            ],
            [
                "versions.yml:md5,07e86bb6f39d315279103919a01645a5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T14:44:51.260709908"
    }
}